### ISSUE TITLE
[BUG] sample_utils samplers return a constant token off the import thread — every mlx_lm.server request at temp > 0

### ISSUE THREAD

**Issue by hanjukim** [ml-explore/mlx-lm#1675]
**Describe the bug**

`sample_utils` compiles five functions with the random state captured at import:

```python
@partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)
def categorical_sampling(logits, temp):
    return mx.random.categorical(logits * (1 / temp))
```

`mx.random.state` is thread-local, and the decoration runs on whichever thread imports the
module — normally the main thread. A call from any *other* thread then reads state arrays that
belong to the tracing thread, and that call never advances them. The function returns the same
index on every step.

`mlx_lm.server` supplies exactly that pairing: it imports on the main thread and generates on
`self._generation_thread` (`server.py:451`). **So every `mlx_lm.server` request with
`temp > 0` samples from a frozen sampler** — `--temp`, `--top-p`, `--top-k` and `--min-p` have
no effect on the result no matter what the command line or request body says, and `seed` cannot
change it either (`mx.random.seed` at `server.py:956` runs on the right thread and still cannot
reach the captured state).

`temp == 0` is unaffected: `make_sampler` returns `mx.argmax` (`sample_utils.py:46`) and the
broken function is never called.

**To Reproduce**

No model, no server:

```python
import threading
from functools import partial
import mlx.core as mx

logits = mx.zeros((1, 4))
mx.eval(logits)                     # main thread touches MLX first

def draw(logits, temp):
    return mx.random.categorical(logits * (1 / temp))

# captured on the main thread, exactly as sample_utils.py does at import
compiled = partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)(draw)

def six(fn):
    mx.random.seed(0)
    return [int(fn(logits, 1.0).item()) for _ in range(6)]

def on_worker(fn):
    box = {}
    t = threading.Thread(target=lambda: box.update(r=six(fn)))
    t.start(); t.join()
    return box["r"]

print("main   compiled      ", six(compiled))
print("worker compiled      ", on_worker(compiled))
print("worker uncompiled    ", on_worker(draw))

def compiled_on_worker():
    fn = partial(mx.compile, inputs=mx.random.state, outputs=mx.random.state)(
        lambda l, t: mx.random.categorical(l * (1 / t)))
    return six(fn)

box = {}
t = threading.Thread(target=lambda: box.update(r=compiled_on_worker()))
t.start(); t.join()
print("worker compiled there", box["r"])
```

Output, identical on three consecutive runs (mlx 0.32.0, mlx-lm 0.31.3, M5 Max, macOS 25.6.0):

```
main   compiled       [0, 2, 3, 1, 2, 3]
worker compiled       [0, 0, 0, 0, 0, 0]
worker uncompiled     [0, 2, 3, 1, 2, 3]
worker compiled there [0, 2, 3, 1, 2, 3]
```

Line 2 is the bug. Lines 3 and 4 localise it: `mx.random.categorical` itself is fine on a
worker, and tracing the same body *on* the calling thread is fine too. The freeze needs both
halves — the capture on one thread and the call on another.

A plain (non-random) captured state does not behave this way: a counter array compiled with
`inputs=`/`outputs=` advances normally on a worker. What singles out `mx.random.state` is that
it is thread-local, so the captured arrays are the wrong thread's rather than merely stale.

**Effect on real generation**

The same `mlx_lm.chat`-shaped call, one process, one seed, `Qwen3.6-35B-A3B-8bit`, temp 1.0,
top-p 0.95, a 1099-token prompt that opens a `<think>` block the model has to close:

| where it runs | closes the block | tokens |
|---|---|---|
| main thread | yes | 38759 |
| worker thread | **no** | 131072 (the whole budget) |
| worker thread, `categorical_sampling` uncompiled | yes | 38759, byte-identical to the main-thread output |

Three seeds on the main thread closed 3/3; the same three on a worker closed 0/3. Sixteen
`mlx_lm.server` rounds of the same model over the same prompt closed 0/16.

**Expected behavior**

A sampler called from a worker thread samples. Failing that, it should raise rather than
silently return a constant index — the current behaviour is a wrong result with no signal.

**Possible fixes**

* Do not capture `mx.random.state` at import time in `sample_utils` — resolve it per call, or
  compile lazily on the thread that first calls the sampler.
* Or drop `mx.compile` from `categorical_sampling`. Replacing that one function with an
  undecorated `lambda logits, temp: mx.random.categorical(logits * (1 / temp))` was enough to
  fix generation in the table above; the other four decorated functions draw no randomness, so
  a frozen state cannot change what they return.

Happy to send a PR for either shape.

**Desktop (please complete the following information):**
 - OS Version: macOS 25.6.0 (Darwin 25.6.0, arm64)
 - Version: mlx 0.32.0, mlx-lm 0.31.3, Python 3.14.6

**Comment by katlun-lgtm** [comment]
Reproduced on an M3 Max (mlx 0.32.0, mlx-lm main), and the mechanism is a little sharper than the report: the stuck token is always draw #1 of the correct sequence for whatever seed the calling thread has, so the thread-local state is being read at call time but never advanced. Across seeds 0/1234/7/99 the frozen token matched the first uncompiled draw every time.

That means one detail is worth correcting: `mx.random.seed` on the generation thread does reach it — it picks which token you freeze on (seed 0 → 7, 1234 → 6, 7 → 3, 99 → 4) — rather than being ignored outright. The user-visible effect is the same, zero diversity.

`temp=0` is unaffected as you say, and all four of temp/top_p/top_k/min_p freeze identically. Building the sampler on the worker thread doesn't help either, since the compiled functions are decorated at import. `MLX_DISABLE_COMPILE=1` makes the worker output byte-identical to the main thread again, which both confirms the compile capture as the cause and gives people running the server a workaround in the meantime.

**Comment by pierre427** [comment]
There is already an open PR for this: **#1593** ("Drop import-level sampler compilation",
open since 2026-07-21). It moves the `mx.compile` from the five import-time decorators into
`make_sampler()`, so the sampler is compiled per call and binds the state of whichever thread
builds it. That keeps the compiled path — unlike `MLX_DISABLE_COMPILE=1`, which restores
correctness by disabling compilation process-wide.

It also lines up with @katlun-lgtm's correction above: with the compile moved into
`make_sampler()`, the calling thread's state is both read *and* advanced, so the draw sequence
progresses instead of freezing on draw #1.

Worth separating from **#1331**, which targets the same user-visible symptom from #1245 but
narrowly: it swaps to uncompiled sampling only when a request is *both* seeded and stochastic,
and bypasses prompt-cache reuse for those requests. That leaves the ordinary unseeded
`temp > 0` server request — the case this issue is about — still sampling from a frozen
sampler, and costs prefix-cache reuse on the requests it does cover. #1593 addresses the
general case with no cache-reuse penalty; it declares `Fixes #1245` / `Closes #1331`.

We've been running #1593 on our fork since 2026-07-23 across long-lived
`mlx_lm.server` processes with no issues. Happy to add a thread-level regression test to that
PR if it would help move it along — #1593 already carries one in `tests/test_sample_utils.py`.

**Comment by zcbenz** [comment]
Closing as a duplicate of https://github.com/ml-explore/mlx-lm/issues/1245.
